servant/src/Servant/API/ResponseHeaders.hs

```diff
@@ -97,24 +98,19 @@ type family HeaderValMap (f :: * -> *) (xs :: [*]) where
 
 class BuildHeadersTo hs where
     buildHeadersTo :: [HTTP.Header] -> HList hs
-    -- ^ Note: if there are multiple occurrences of a header in the argument,
-    -- the values are interspersed with commas before deserialization (see
-    -- <http://www.w3.org/Protocols/rfc2616/rfc2616-sec4.html#sec4.2 RFC2616 Sec 4.2>)
 
 instance {-# OVERLAPPING #-} BuildHeadersTo '[] where
     buildHeadersTo _ = HNil
 
 instance {-# OVERLAPPABLE #-} ( FromHttpApiData v, BuildHeadersTo xs, KnownSymbol h )
          => BuildHeadersTo (Header h v ': xs) where
-    buildHeadersTo headers =
-      let wantedHeader = CI.mk . pack $ symbolVal (Proxy :: Proxy h)
-          matching = snd <$> filter (\(h, _) -> h == wantedHeader) headers
-      in case matching of
-        [] -> MissingHeader `HCons` buildHeadersTo headers
-        xs -> case parseHeader (BS.init $ BS.unlines xs) of
-          Left _err -> UndecodableHeader (BS.init $ BS.unlines xs)
-             `HCons` buildHeadersTo headers
-          Right h   -> Header h `HCons` buildHeadersTo headers
+    buildHeadersTo headers = case L.find wantedHeader headers of
+      Nothing -> MissingHeader `HCons` buildHeadersTo headers
+      Just header@(_, val) -> case parseHeader val of
+        Left _err -> UndecodableHeader val `HCons` buildHeadersTo (L.delete header headers)
+        Right h   -> Header h `HCons` buildHeadersTo (L.delete header headers)
+      where wantedHeader (h, _) = h == wantedHeaderName
+            wantedHeaderName = CI.mk . pack $ symbolVal (Proxy :: Proxy h)
```

Comment:
I had a look at multiple RFCs regarding HTTP header fields, i.e.:
- [1] https://www.rfc-editor.org/rfc/rfc7230#section-3.2.2
- [2] https://www.rfc-editor.org/rfc/rfc9110#name-field-order
- [3] https://www.rfc-editor.org/rfc/rfc6265.html
- [4] https://www.rfc-editor.org/rfc/rfc8941.html
- [5] https://httpwg.org/http-extensions/draft-ietf-httpbis-retrofit.html

At the end, we do not still have a precise definition for which field lines with the same field name are allowed to be present in the HTTP message or not (except for the `Set-Cookie`), and which ones are allowed to be _merged_ in some way.
We have something about it, see [4] and [5], and those RFCs are pretty new or are still drafts.

I think that the current (fixed) approach of not manipulating the _requested_ header fields is ok.
We'll be able to create also _not valid_ messages, and this is good for testing (here being _not valid_ means that we could generate HTTP messages with header fields with the same field name, for which we require to have just one line, and not multiple ones).

I would just add a short comment that summarise this behaviour, like: _The current implementation do not manipulate HTTP header field lines in any way, like merging field lines with the same field name in a single line._

Reply:
Thanks for doing the research @marinelli! I've added the suggested comment.